mamedov_timur/printing.cpp

```diff
@@ -35,27 +37,32 @@ std::string print_edge(const Graph::Edge& edge, const Graph& graph) {
 std::string print_graph(const Graph& graph) {
   std::stringstream json_string;
   json_string << "{\n\t\"vertices\": [";
-  const auto vertices = graph.get_vertices();
-  const auto edges = graph.get_edges();
+  const auto& vertices = graph.get_vertices();
+  const auto& edges = graph.get_edges();
 
-  auto it_vertices = vertices.begin(), prev_it_vertices = it_vertices;
-  ++it_vertices;
+  if (!vertices.empty()) {
+    auto it_vertices = vertices.begin();
+    auto prev_it_vertices = it_vertices;
+    ++it_vertices;
 
-  for (; it_vertices != vertices.end(); ++it_vertices, ++prev_it_vertices) {
-    json_string << print_vertex(prev_it_vertices->second, graph) << ",";
+    for (; it_vertices != vertices.end(); ++it_vertices, ++prev_it_vertices) {
+      json_string << print_vertex(prev_it_vertices->second, graph) << ",";
+    }
+    json_string << print_vertex(prev_it_vertices->second, graph);
   }
```

Comment:
Писать неполный `for` считается плохой практикой. Выносить переменные во внешнюю зону видимости считается плохой практикой. Лучше упростить, как-нибудь, например так:
```suggestion
  if (!vertices.empty()) {
    bool is_first_iteration = true;
    for (auto it = vertices.cbegin(); it != vertices.cend(); ++it) {
      if (!is_first_iteration) {
        json_string << ",";
      }
      json_string << print_vertex(it->second, graph);
      is_first_iteration = false;
    }
  }
```